docs/custom-components-pkg.qmd

```diff
@@ -0,0 +1,405 @@
+
+
+### `tsconfig.json`
+This file is used to configure typescript, which we are using to write our component. Like the `pyproject.toml` file, a deep dive into the structure of this file is outside the scope of this article, but you can find more information in the [typescript docs.](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html)
+
+
+### `example-app/app.py`
+
+```{.python}
+from fancy_color_picker import fancy_color_picker
+
+from shiny import App, render, ui
+
+app_ui = ui.page_fluid(
+    fancy_color_picker("myComponent"),
+    ui.output_text("valueOut"),
+)
+
+def server(input, output, session):
+    @render.text
+    def valueOut():
+        return f"Value from input is {input.myComponent()}"
+
+app = App(app_ui, server)
+```
+
+
+This is a simple example app that can be used to test the component while developing. It uses the `fancy_color_picker` function we defined in `fancy_color_picker.py` to add the component to the app. It also uses the `render.text` decorator to render the value of the input to the page.
+
+![Example app running with color picker](custom-components-pkg-example-app.png)
+
+::: {.callout-tip collapse="true"}
+## What about an output binding?
+
+In our output binding example we defined an output that conveniently _displays_ colors. If we were packaging up two components like this we could/should modify the example app to showcase both of them.
+
+```{.python filename="app.py"}
+from fancy_color_picker import fancy_color_picker, output_color, render_color
+
+from shiny import App, ui
+
+app_ui = ui.page_fluid(
+    fancy_color_picker("myComponent"),
+    output_color("myColor"),
+)
+
+def server(input, output, session):
+    @render_color
+    def myColor():
+        return input.myComponent()
+
+app = App(app_ui, server)
+```
+
+:::
+
+
+
+## Development workflow
+
+
+Now that we’ve seen the files that make up the package, let’s talk about how to develop the component.
+
+While there are lots of ways to develop components with live-feedback (e.g. Storybook, dev servers like `vite`, etc) an easy way to develop a component with our package structure is to use the example app, an editable mode pip install, and the watch mode for our build step. We can do this with the following steps:
+
+1. Install the package in [“editable mode”](https://setuptools.pypa.io/en/latest/userguide/development_mode.html) with `pip install -e .`
+
+2. Run the bundler in watch mode with `npm run watch`. This will watch the `srcts` directory for changes and automatically rebuild the javascript when it detects a change.
+
+3. Run the example app in live-reload mode. If you're using VScode, the [Shiny for Python extension](https://marketplace.visualstudio.com/items?itemName=Posit.shiny-python) enables this automatically when pressing the run button above the app script.
+
+Now you can update your component javascript/python functions and your app will automatically reload with the changes. Happy developing!
```

Comment:
I would but Pypi has frozen new account development since I started this article so I don't actually have the ability to test it out. I'll put a link about how to do it, though.

Reply:
Added link out to the Python Packaging User Guide which honestly is most likely a much better resource than I could have hoped to write given how terse it already is.